Documentation/replacing-component.md

```diff
@@ -0,0 +1,266 @@
+
+
+One of the additional side effects of switching to a more fine-grained protocol structure is that we can use the type system to express some of the requirements a web service must satisfy. For example, we should be able to define the protocols/types/builder in a way that ill-formed web services will cause a compile-time error.
+
+
+
+
+### Issues caused by the current implementation
+
+
+#### Invalid Components
+
+Both of these components currently compile, despite not being valid components:
+
+```swift
+// empty component which implements neither `content` nor `handle()`
+struct A: Component {}
+
+// component which implements both `content` and `handle()`
+struct B: Component {
+    var content: some Component { ... }
+    func handle() -> Response { ... }
+}
+```
+
+More generally speaking, fully half of the four possible combinations of conforming to the `Component` protocol will result in invalid components:
+
+<!--- Implement neither `content` nor `handle()`: ❌
+- Implement only `content`: ✅
+- Implement only `handle()`: ✅
+- Implement both `content` and `handle()`: ❌-->
+
+
+|            | `content` | `handle()` | component valid? |
+| :--------- | :-------: | :--------: | :--------------: |
+| implements |  no       | no         | ❌       |
+| implements |  yes      | no         | ✅               |
+| implements |  no       | yes        | ✅               |
+| implements |  yes      | yes        | ❌               |
+
+
+
+
+#### Type inference
+
+Because the current implementation defaults both `Content` and `Response` to `Never`, you can end up in situations where a seemingly properly defined component is in fact ill-formed.
+
+
+Consider the following example:
+
+```swift
+struct User: Codable {
+    let id: ID
+    let name: String
+    let age: Int
+}
+
+struct GetUser: Component {
+    @Parameter("userId")
+    var userId: User.ID
+    
+    func handle() -> User {
+        // ...
+    }
+}
+```
+
+
+This component, despite compiling just fine and not causing a single warning or error, will never work as intended.  
+The reason for this is that `User` doesn't conform to `ResponseEncodable`, meaning that swiftc will resolve `GetUser.Response` to `Never` (the default value), which will cause Apodini to invoke the `handle` function implemented in the `Response == Never` protocol extension.  
+The result is that, when a request is sent to the component's endpoint, it will crash at runtime.
+
+
+**Generally speaking, if we can use the type system to statically detect and reject ill-formed programs, instead of having to perform these checks at runtime, we should always do so.**
+
+
+
+## Proposed replacement
+
+
+### The new protocol structure
+
+My current implementation in [experimental/dsl-node-protocols][branch_url] replaces `Component` with the following two protocols:
+
+```swift
+protocol EndpointNode {
```

Comment:
I originally suggested in https://github.com/Apodini/Apodini/pull/26#discussion_r538396896 that the type providing the handle function also conforms to the type building the static structure of the web service (`EndpointProvidingNode`) to enable a clear hierarchy of protocols. Out of curiosity, what would be the advantage of using a completely separate protocol with no connection in this case?

Reply:
My idea here was that having them be separate would prevent accidentally passing one where we expect the other.
Currently, you can use "components with a handle function" and "components with a content property" interchangeably, this wouldn't be allowed anymore